### DISCUSSION TITLE
Running Docker Component Locally Which Relies on GreenGrass Core IPC Client

### DISCUSSION THREAD

**Discussion by HieronymusLex** [aws/aws-iot-device-sdk-js-v2#537] [Q&A] [answered]
We have a NodeJS component which runs fine on a GreenGrass core (Raspberry Pi) as a Docker container. The component uses the `greengrasscoreipc.Client` to subscribe and publish messages to 2 topics on the core. We want to be able to run the component locally on an OSX laptop (containerised or not doesn't matter locally, either works) during dev to facilitate hot reloading and rapid feedback. Is there a recommended approach for this? 

Currently when we start the component locally, the `createClient` throws `RpcError: Eventstream RPC client configuration must have a valid host name` which is expected since it's not running on a core.

What would be the recommended setup for local dev of a component like this?

**Comment by bretambrose** [comment, marked as answer]
Update: I've been pointed at https://youtu.be/adx3zUGRoWw?t=303&si=CcOgHoGp7qvwdwUp as a possible solution to your question.

My original reply (which is a little incomplete on information):

If you want to run a component (or any application that uses GG IPC for that matter) without it being managed by a core you'll need to setup some specific environment variables (this is how I run GG IPC directly from an IDE for that matter, I don't actually manage components):

https://github.com/aws/aws-iot-device-sdk-js-v2/blob/main/lib/greengrasscoreipc.ts#L23-L24

* AWS_GG_NUCLEUS_DOMAIN_SOCKET_FILEPATH_FOR_COMPONENT - this is always the same value but I forget what it is off the top of my head 
* SVCUID - This one is more awkward and requires a manual step.  This variable is an opaque token/"secret" that changes every time the nucleus starts.  I set up my nucleus with a trivial component that logs this value to the GG service logs.  When I need to run an IPC script from the debugger, I check the logs to see what the current value is and set the environment variable appropriately.

More information on these can be found here: https://docs.aws.amazon.com/greengrass/v2/developerguide/component-environment-variables.html

TLDR: Make a trivial component that logs `AWS_GG_NUCLEUS_DOMAIN_SOCKET_FILEPATH_FOR_COMPONENT` and `SVCUID` environment values to the GG logs.  Run your IPC script from an IDE with those environment variables set to the same values as what appears in the GG logs.
